### PR TITLE
Added NumericRange filter which gives the possibility of filtering …

### PR DESCRIPTION
**Pull request by christopherhero**
| Q               | A
| --------------- | -----
| Branch?         | 1.13
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | #
| License         | MIT

While working on adding a custom filter, I thought that it could be in the bundle.
I add `range numerical` filter which can filter on numerical fields in range (from-to) Like existing DateFilter.

Also in this pull request is a small change in two classes:
- `Doctrine\ORM\ExpressionBuilder`
- `Doctrine\ORM\DataSource`.

I suggested adding filtering capability also when using HAVING instead of WHERE in the query. For example, when we use `count (*)`. In current architecture is hard to distinguish the type of condition (WHERE or HAVING). Another way than suggested by me is to add pubilic methods e.g `havingLessThan` to `ORM\ExpressionBuilder` and new interface "HavingExpressionBuilder"

Please give your opinion what do you think about it?

This idea also came while working on this filter because I had to filter in queries with the `HAVING` clause.

### CODE REVIEW

**Comment by Roshyo** [comment]
Hi @christopherhero and thank you for your PR.

But I feel this one is a bit too complex for the value it should bring.
I would also love to have a numeric filter out of the box. But IMO it should just look like the current `MoneyFilter` without the currency part. What do you think ?

**Comment by christopherhero** [comment]
@Roshyo Hi thanks :)


I have lightly changed `RangeNumericalFiolter` to `NumericalFIlter` (I also removed the logic for the expression builder, I will add that in a separate pr)

**Comment by christopherhero** [comment]
@Roshyo what you think now :) ?

**Comment by Roshyo** [comment]
LGTM. Let's wait for another approval before merging

**Comment by Roshyo** [comment]
@christopherhero Could you try to rebase on 1.13?

**Comment by christopherhero** [comment]
@Roshyo updated

**Review by Roshyo — APPROVED** [review]

File: src/Component/Filter/NumericalFilter.php
Comment (vvasiloi) [inline src/Component/Filter/NumericalFilter.php]: Let's say I want to get all the variants that have on_hold > 5, do I have to write 0.05 in the input?

Comment (christopherhero) [reply src/Component/Filter/NumericalFilter.php]: No, in this case you have to configure `scale: 0` in your grid config e.g

```
sylius_grid:
    grids:
        ...
            filters:
                numberField:
                    type: numerical
                    form_options:
                        scale: 0
```

**Comment by diimpp** [comment]
+1 for adding numeric filter, current list looks lonely without it :smiley: 

I've a naming concern, it feels more like `RangeFilter`, rather than `NumericFilter`. 

How about to synchronize naming and capabilities with api-platform?
https://api-platform.com/docs/core/filters/#numeric-filter `<eq>`
https://api-platform.com/docs/core/filters/#range-filter `<lt|gt|lte|gte|between>`

File: src/Component/Filter/NumericalFilter.php
Comment (vvasiloi) [reply src/Component/Filter/NumericalFilter.php]: Cool, thanks! Maybe make 0 the default then?

**Comment by christopherhero** [comment]
@diimpp Thanks for review :) 
Regarding Api platfrom, could you give some example of what you mean ? 
Because I don't fully understand, thanks :)

**Comment by christopherhero** [comment]
@diimpp What else to change :) ?

File: src/Bundle/Form/Type/Filter/RangeNumericFilterType.php
Comment (diimpp) [inline src/Bundle/Form/Type/Filter/RangeNumericFilterType.php]: Either
```suggestion
final class NumericRangeFilterType extends AbstractType
```
or

```suggestion
final class RangeFilterType extends AbstractType
```

@loic425 Hi, wdyt?

File: src/Bundle/Form/Type/Filter/RangeNumericFilterType.php
Comment (diimpp) [inline src/Bundle/Form/Type/Filter/RangeNumericFilterType.php]: Scale depends on rounding mode, so it should be possible to configure as well.

image

https://symfony.com/doc/current/reference/forms/types/number.html#rounding-mode

File: src/Bundle/Form/Type/Filter/RangeNumericFilterType.php
Comment (diimpp) [inline src/Bundle/Form/Type/Filter/RangeNumericFilterType.php]: Same

File: src/Bundle/Resources/config/services/filters.xml
Comment (diimpp) [inline src/Bundle/Resources/config/services/filters.xml]:
```suggestion
            <tag name="sylius.grid_filter" type="numeric_range" form-type="Sylius\Bundle\GridBundle\Form\Type\Filter\NumericRangeFilterType" />
```
```suggestion
            <tag name="sylius.grid_filter" type="range" form-type="Sylius\Bundle\GridBundle\Form\Type\Filter\RangeFilterType" />
```

File: src/Bundle/Form/Type/Filter/RangeNumericFilterType.php
Comment (diimpp) [inline src/Bundle/Form/Type/Filter/RangeNumericFilterType.php]:
```suggestion
        return 'sylius_grid_filter_numeric_range';
```
```suggestion
        return 'sylius_grid_filter_range';
```

File: src/Bundle/Resources/config/services/filters.xml
Comment (diimpp) [inline src/Bundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="sylius.grid_filter.numeric_range" alias="Sylius\Component\Grid\Filter\NumericRangeFilter" />
```
```suggestion
        <service id="sylius.grid_filter.range" alias="Sylius\Component\Grid\Filter\RangeFilter" />
```

File: src/Bundle/Resources/config/services/filters.xml
Comment (diimpp) [inline src/Bundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="Sylius\Bundle\GridBundle\Form\Type\Filter\NumericRangeFilterType">
```
```suggestion
        <service id="Sylius\Bundle\GridBundle\Form\Type\Filter\RangeFilterType">
```

File: src/Bundle/Resources/config/services/filters.xml
Comment (diimpp) [inline src/Bundle/Resources/config/services/filters.xml]:
```suggestion
        <service id="sylius.form.type.grid_filter.numeric_range" alias="Sylius\Bundle\GridBundle\Form\Type\Filter\NumericRangeFilterType" />
```
```suggestion
        <service id="sylius.form.type.grid_filter.range" alias="Sylius\Bundle\GridBundle\Form\Type\Filter\RangeFilterType" />
```

File: src/Component/Filter/RangeNumericFilter.php
Comment (diimpp) [inline src/Component/Filter/RangeNumericFilter.php]:
```suggestion
final class NumericRangeFilter implements FilterInterface
```
```suggestion
final class RangeFilter implements FilterInterface
```

File: src/Component/Filter/RangeNumericFilter.php
Comment (diimpp) [inline src/Component/Filter/RangeNumericFilter.php]: I assume if I will write `field: 1`, then it would be int and we need string. `Date` filter do this conversion as well.
```suggestion
        $field = (string) $options['field'] ?? $name;
```

File: src/Component/Filter/RangeNumericFilter.php
Comment (diimpp) [inline src/Component/Filter/RangeNumericFilter.php]: I wonder which default scale would be sensible.

@Sylius/development-team wdyt?

SF numeric form type uses `3`.

File: src/Component/Filter/RangeNumericFilter.php
Comment (diimpp) [inline src/Component/Filter/RangeNumericFilter.php]: Since phpstan now understands more complex docblocks, something like this should be done.

```suggestion
    /** @param array<array-key, string> $data */
```

**Review by diimpp — COMMENTED** [review]
Hi @christopherhero, sorry for late reply. :pray: 

Please see the comments.

Filter should allow to specify comparison strictness, like `lt|gt|lte|gte`, e.g. `greater than` and `greater than or equel`

**Comment by christopherhero** [comment]
@diimpp

**Comment by christopherhero** [comment]
what do you think about it now ;) 
@diimpp @vvasiloi

File: src/Component/spec/Filter/NumericRangeFilterSpec.php
Comment (vvasiloi) [inline src/Component/spec/Filter/NumericRangeFilterSpec.php:29]: This should check that `$dataSource->restrict()` was not called, like in the ExistsFilter for example: https://github.com/Sylius/SyliusGridBundle/blob/5d3f702947f17502e0d6e173276086c1d8ebcf9b/src/Component/spec/Filter/ExistsFilterSpec.php#L30-L35

For some reason, the same was done for the MoneyFilter, but it does not seem right:
https://github.com/Sylius/SyliusGridBundle/blob/5d3f702947f17502e0d6e173276086c1d8ebcf9b/src/Component/spec/Filter/MoneyFilterSpec.php#L34-L42

/cc @Sylius/core-team

File: src/Component/spec/Filter/NumericRangeFilterSpec.php
Comment (vvasiloi) [inline src/Component/spec/Filter/NumericRangeFilterSpec.php]: Does this work? The second argument should be an integer. 🤔

File: src/Component/spec/Filter/NumericRangeFilterSpec.php
Comment (vvasiloi) [inline src/Component/spec/Filter/NumericRangeFilterSpec.php:168]: Maybe include a test where the scale is not 0, since that's the default value now.

**Review by vvasiloi — CHANGES_REQUESTED** [review]

File: src/Component/spec/Filter/NumericRangeFilterSpec.php
Comment (christopherhero) [reply src/Component/spec/Filter/NumericRangeFilterSpec.php:42]: example -> `vendor/sylius/grid-bundle/src/Bundle/spec/Doctrine/PHPCRODM/ExpressionBuilderSpec.php:78`

Comment (christopherhero) [reply src/Component/spec/Filter/NumericRangeFilterSpec.php]: Yes, works (but I changed)

**Comment by christopherhero** [comment]
@vvasiloi

File: src/Component/Filter/NumericRangeFilter.php
Comment (vvasiloi) [inline src/Component/Filter/NumericRangeFilter.php:45]: This statement can be moved up, outside the if statement, and the one on line 57 can be removed.

File: src/Component/Filter/NumericRangeFilter.php
Comment (vvasiloi) [inline src/Component/Filter/NumericRangeFilter.php]: Inclusive is always boolean, so this can be simplified to `if ($inclusive)`. Same for the other one at line 60.

**Review by vvasiloi — APPROVED** [review]
Only 2 more observations I missed in the previous review, but it can be merged even without addressing those.

File: src/Component/Filter/RangeNumericFilter.php
Comment (vvasiloi) [reply src/Component/Filter/RangeNumericFilter.php]: I think 0 is the most sensible. I don't think filtering with decimal precision is more common than with whole numbers.

File: src/Component/Filter/NumericRangeFilter.php
Comment (christopherhero) [reply src/Component/Filter/NumericRangeFilter.php:45]: Unfortunately, but it cannot be removed due to the fact that they are two different fields `inclusive_from` and `inclusive_to`.

Therefore, moving this `statement` higher in this case will not make sense.
Or did I miss something :) ?

Comment (vvasiloi) [reply src/Component/Filter/NumericRangeFilter.php:45]: Right, my mistake.

**Review by oallain — APPROVED** [review]

**Review by NoResponseMate — APPROVED** [review]

**Comment by NoResponseMate** [comment]
Thank you, @christopherhero!

**Comment by vvasiloi** [comment]
Thank you @christopherhero, for having patience and finishing the work even after so much time passed! 🥇

**Pull request merged**
